Customize host root volume size by Terrafrom provisioning

r0b2g1t opened the pull request:
The changes allow the volume size of the different nodes to be adjusted during deployment with Terraform and increase flexibility. The following variables are available in the ```cluster.tf```.
```
Bastion    - bastion_root_volume_size_in_gb
Etcd       - etcd_root_volume_size_in_gb
GlusterFS  - gfs_root_volume_size_in_gb
Master     - master_root_volume_size_in_gb
Worker     - node_root_volume_size_in_gb
```
The default value is set to 10GB in ```variables.tf``` of Terraform.

r0b2g1t commented:
> @r0b2g1t thank you for your contribution, as it stands I think that this is not backwards compatible. Can we switch to a default volume size of 0 and the cinder volume part only if volume_size > 0 ?

I have check the code to switch on/off the ```block_device``` part with Terraform. The only solution I see it to duplicate the node ressoruces code because it is not reusable.

Miouge1 commented:
Terraform 0.12 will introduce if/else etc... so maybe it will be easier once release, until then only a few tricks are available like count with ternary.

@holmsten maybe you know more about this?

holmsten commented:
> Terraform 0.12 will introduce if/else etc... so maybe it will be easier once release, until then only a few tricks are available like count with ternary.
> 
> @holmsten maybe you know more about this?

We have discussed this in-house at ELASTX before actually. I think you and me think alike. I don't want to force users to use volumes. I see two ways forward.

1. Wait for if/else block support in Terraform.
2. Create new resources for supporting VMs with block devices, i.e. something like `openstack_compute_instance_v2.k8s_master_boot_volume` and then define with a count in cluster.tf: `number_of_k8s_masters_boot_volume = 3`

Actually the second option is what we've done in-house while waiting for Terraform 0.12. I can create a PR for it if we want to compare and evaluate.

r0b2g1t commented:
@holmsten the PR would be nice to evaluate it.

Miouge1 commented:
FYI TF 0.12 is planned for release in Q1 2019

r0b2g1t commented:
I've added a new commit. It give the possibilty to use instances with or without block_device. By default the ```volume_size```-values are set to ```0``` which disable the block_device use.

r0b2g1t commented:
I fixed an error with the hosts script. Which parse the image-name of the instance resources from the ```tfstate``` inventory file.

r0b2g1t commented:
@Miouge1 I have integrate all changes and your recommendation.

Miouge1 commented:
@r0b2g1t sorry for the delay, we've been working on adding CI to contrib/terraform/openstack. We think that the addition of automatic tests will help review those PRs faster.

You can see that `tv-validate-openstack` is complaining about a couple of formatting points.

r0b2g1t commented:
> @r0b2g1t sorry for the delay, we've been working on adding CI to contrib/terraform/openstack. We think that the addition of automatic tests will help review those PRs faster.
> 
> You can see that `tf-validate-openstack` is complaining about a couple of formatting points.

No Problem, I have fixed the formatting.

Miouge1 commented:
ci check this

Miouge1 commented:
@r0b2g1t can you rebase this on latest master? we've added a `tf-ovh` job that allows us to test terraform on OpenStack provided by OVH. I think they have cinder so we could see this in action.

r0b2g1t commented:
> @r0b2g1t can you rebase this on latest master? we've added a `tf-ovh` job that allows us to test terraform on OpenStack provided by OVH. I think they have cinder so we could see this in action.

Done :-)

rptaylor commented on contrib/terraform/openstack/modules/compute/main.tf:
```diff
@@ -167,9 +249,46 @@ resource "openstack_compute_instance_v2" "k8s_master_no_etcd" {
   }
 }
 
+resource "openstack_compute_instance_v2" "k8s_master_no_etcd_custom_volum_size" {
```
Looks like a typo, should be k8s_master_no_etcd_custom_volume_size  (same with the 2 other instances of this variable).

rptaylor commented:
It would be very useful to be able to specify volume sizes. 
I already privately tweaked my Kubespray deployment to support this, would be great to have upstream.
@r0b2g1t  looks like the automated build test is blocked on a yaml lint check and there is a complaint about trailing whitespace on this line:
```.git/rebase-apply/patch:12: trailing whitespace.
  node_root_volume_size_in_gb                  = "${var.node_root_volume_size_in_gb}" ```
Other than that looks like it needs another rebase ... and then a ping to Miouge1 probably if the checks are successful.

r0b2g1t commented:
I'm still interessted in this PR and  I will resolve the merge conflicts.

r0b2g1t commented:
> It would be very useful to be able to specify volume sizes.
> I already privately tweaked my Kubespray deployment to support this, would be great to have upstream.
> @r0b2g1t looks like the automated build test is blocked on a yaml lint check and there is a complaint about trailing whitespace on this line:
> 
> ```
> .git/rebase-apply/patch:12: trailing whitespace.
>   node_root_volume_size_in_gb                  = "${var.node_root_volume_size_in_gb}" 
> ```
> 
> Other than that looks like it needs another rebase ... and then a ping to Miouge1 probably if the checks are successful.

I merged the kubespray master in the volume_size branch and fixed the typos but I can't find the whitespace error.

rptaylor commented:
@r0b2g1t The output of the lint check is here:
https://gitlab.com/kargo-ci/kubernetes-sigs-kubespray/-/jobs/283209995

Hmm, the line it is complaining about is actually a .tf file, not YAML. I think that is just a warning and not relevant. The actual problem is shown later in the output:
```
Applying: disable block_device by default value
error: Failed to merge in the changes.
Using index info to reconstruct a base tree...
M	contrib/terraform/openstack/modules/compute/main.tf
M	contrib/terraform/openstack/variables.tf
Falling back to patching base and 3-way merge...
Auto-merging contrib/terraform/openstack/variables.tf
Auto-merging contrib/terraform/openstack/modules/compute/main.tf
CONFLICT (content): Merge conflict in contrib/terraform/openstack/modules/compute/main.tf
Patch failed at 0003 disable block_device by default value
```
One of the commits seems to introduce a merge conflict. I think this automated lint test uses a rebasing script which may run into trouble easily. I recall this happened in one of my PRs and I just rebased and squashed all the commits together which fixed it, so maybe that would work for you.

r0b2g1t commented:
> @r0b2g1t The output of the lint check is here:
> https://gitlab.com/kargo-ci/kubernetes-sigs-kubespray/-/jobs/283209995
> 
> Hmm, the line it is complaining about is actually a .tf file, not YAML. I think that is just a warning and not relevant. The actual problem is shown later in the output:
> 
> ```
> Applying: disable block_device by default value
> error: Failed to merge in the changes.
> Using index info to reconstruct a base tree...
> M	contrib/terraform/openstack/modules/compute/main.tf
> M	contrib/terraform/openstack/variables.tf
> Falling back to patching base and 3-way merge...
> Auto-merging contrib/terraform/openstack/variables.tf
> Auto-merging contrib/terraform/openstack/modules/compute/main.tf
> CONFLICT (content): Merge conflict in contrib/terraform/openstack/modules/compute/main.tf
> Patch failed at 0003 disable block_device by default value
> ```
> 
> One of the commits seems to introduce a merge conflict. I think this automated lint test uses a rebasing script which may run into trouble easily. I recall this happened in one of my PRs and I just rebased and squashed all the commits together which fixed it, so maybe that would work for you.

To rebase and put all together again, this sounds like a workaround to eliminate the symptoms instead of the cause in the pipeline script.

@Miouge1 or @mirwan: Maybe someone can look at the pipeline and fix the problem?

rptaylor commented:
@r0b2g1t When I rebased and squashed my commits it only took ~ 2 minutes, but I think I took a different approach (squash my commits while rebasing my branch onto latest master) whereas it looks like (based on `git log --graph --decorate --pretty=oneline --abbrev-commit`) you rebased master onto your branch? I am not sure, not really a git expert. Sorry I can't help more.

r0b2g1t commented:
> @r0b2g1t When I rebased and squashed my commits it only took ~ 2 minutes, but I think I took a different approach (squash my commits while rebasing my branch onto latest master) whereas it looks like (based on `git log --graph --decorate --pretty=oneline --abbrev-commit`) you rebased master onto your branch? I am not sure, not really a git expert. Sorry I can't help more. Or maybe there really is a merge conflict somewhere.

Rebase is done. Also fixed a small bug.

r0b2g1t commented:
@rptaylor could you check the pull request if everything looks fine can you merge it?

rptaylor commented:
@r0b2g1t The tests are successful now but I don't have merge permission. 
@Miouge1 @mirwan  can you please take a look while this is fresh?

/assign @miouge1

rptaylor commented:
@mirwan This PR should be ready to go, can you review while it's still fresh? Thanks!

Miouge1 commented:
I have tested this, works well. Thank you @r0b2g1t !

/approve

Miouge1 commented:
And LGTM tag missing:

/lgtm
